### PR TITLE
Questions in TAO are scored out of 3, instead 1

### PR DESCRIPTION
**Pull request by pooja31chaudhary**
Questions in TAO are scored out of 3, instead 1 fixed. For multiple response type questions TAO will assign seperate score (which has been assigned to question) for each response.

### CODE REVIEW

**Comment by ttton** [comment]
Hi @pooja31chaudhary, can you please give this pull request a short, clear and meaningful title? Thanks.

**Comment by jack-lin-lrn** [comment]
Hi @pooja31chaudhary,

Can you please give us more detailed information about what the new changes in this PR are for.
It seems to me that some parts doesn’t relate to this PR.
Can you please separate PRs for anything not related to the topic of the PR - i.e. scoring for `clozedropdown` (or ther `cloze` question types)

The issue is still there, here is how I replicate it:
1. Download attachment from `https://learnosity.atlassian.net/browse/WBL-59` and move it to learnosity-qti/data/input/WBL-59
2. Open `data/input/WBL-59/items/29ab95d89f40848bb1cc73d9adacb624.json`
3. Add one more set of response into fields `possible_responses`, `template` and `validation.valid_response.value`.
4. Convert it to qti and back to learnosity by executing commands below:
    1. cd learnosity-qti
    2. ./bin/mo convert:to:qti --input data/input/WBL-59 --output data/output/WBL-59
    3. ./bin/mo convert:to:learnosity --input data/output/WBL-59 --output data/input/WBL-59_BackToLRN --organisation_id 1 
    4. cat data/input/WBL-59_BackToLRN/raw/raw.json | python -m json.tool > data/input/WBL-59_BackToLRN/raw/formatted.json
5. Open `data/input/WBL-59_BackToLRN/raw/formatted.json` 
6. Search `0fb74b46-1f2d-4223-ac0b-d5249f6e626b`
7. You’ll find that score is based on number of responses. (For example, 3 responses are given score 3 instead of score 1 as expected, see below)

```
"type": "clozedropdown",
"validation": {
    "scoring_type": "exactMatch",
    "valid_response": {
        "score": 3,                 <— here
        "value": [
            "Choice A",
            "Choice B",
            "Choice B"
        ]
    }
}
```

Thanks!

**Comment by pooja31chaudhary** [comment]
Hi @jack-lin-lrn , this issue was for conversion of Learnosity:to:qti not for vice versa. You need to process the conversion of attached item data to qti and then upload that converted QTI item to TAO (testing tool) and check there it is scoring out of 3 or 1. It should score out of 1.

**Comment by jack-lin-lrn** [comment]
Hi @pooja31chaudhary ,
We recognised that this PR is not related to conversion from qti to learnosity.
We verified your fixes and it looked good.
However, we don't fully understand the extent of the changes. It would be great if you can add comments for code and clarify some of their behaviours.
In addition, we'd like to ensure that this PR is only for the bug. If there is some code out of scope of the bug, please separate it into different PRs. If the code can’t be separated for any reason, please add clear explanations in the PR description and include a description of any changes that aren’t part of the original scope of the bug. 
Thanks!

**Comment by pooja31chaudhary** [comment]
Hi @jack-lin-lrn , all the changes committed in this PR are related to the same bug. I have added comment where I think you must have not get the reason of changes. Actually this bug is generated because of wrong response processing data. I have made these changes so that it create correct response processing for each question type.

File: src/Processors/QtiV2/Out/QuestionTypes/LongtextMapper.php [hunk 1]
```diff
@@ -13,6 +14,15 @@ public function convert(BaseQuestionType $questionType, $interactionIdentifier,
     {
         /** @var longtext $question */
         $question = $questionType;
+        $questionData = $question->to_array();
+        
+		// $feedbackOptions is used to create feedback or modelfeedback elements in responseProcessing
```
Comment (jack-lin-lrn) [inline src/Processors/QtiV2/Out/QuestionTypes/LongtextMapper.php]: Can you please fix indentation for the comment?

File: src/Processors/QtiV2/Out/QuestionTypes/LongtextV2Mapper.php [hunk 2]
```diff
@@ -13,19 +14,33 @@ public function convert(BaseQuestionType $questionType, $interactionIdentifier,
     {
         /** @var longtextV2 $question */
         $question = $questionType;
-
+        $questionData = $question->to_array();
+        
+		// $feedbackOptions is used to create feedback or modelfeedback elements in responseProcessing
```
Comment (jack-lin-lrn) [inline src/Processors/QtiV2/Out/QuestionTypes/LongtextV2Mapper.php]: and here

File: src/Processors/QtiV2/Out/QuestionTypes/PlaintextMapper.php [hunk 3]
```diff
@@ -13,19 +14,36 @@ public function convert(BaseQuestionType $questionType, $interactionIdentifier,
     {
         /** @var plaintext $question */
         $question = $questionType;
-
+        $questionData = $question->to_array();
+        
+		// $feedbackOptions is used to create feedback or modelfeedback elements in responseProcessing
```
Comment (jack-lin-lrn) [inline src/Processors/QtiV2/Out/QuestionTypes/PlaintextMapper.php]: and here

File: src/Processors/QtiV2/Out/ResponseProcessing/QtiResponseProcessingBuilder.php [hunk 4]
```diff
@@ -19,40 +21,134 @@
...
+            $responseIfScoreComponent = new SetOutcomeValue('SCORE', new BaseValue(BaseType::FLOAT, $score));
+        }
+
+        $responseIfRuleCollection = new ResponseRuleCollection();
+        $responseIfRuleCollection->attach($responseIfScoreComponent);
+
+        // generating response else condition
+        $responseElseRuleCollection = new ResponseRuleCollection();
+		
+		// creating feedback outcome
```
Comment (jack-lin-lrn) [inline src/Processors/QtiV2/Out/ResponseProcessing/QtiResponseProcessingBuilder.php]: and here

File: src/Processors/QtiV2/Out/Validation/PlaintextValidationBuilder.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        /** @var plainText_validation $validation */
+        $responseDeclaration = new ResponseDeclaration($responseIdentifier);
+
+        $responseDeclaration->setCardinality(Cardinality::SINGLE);
+        $responseDeclaration->setBaseType(BaseType::STRING);
+
+        return $responseDeclaration;
+    }
+	
+	public function buildValidation($responseIdentifier, $validation, $isCaseSensitive = true, $distractorRationaleResponseLevel = array())
```
Comment (jack-lin-lrn) [inline src/Processors/QtiV2/Out/Validation/PlaintextValidationBuilder.php]: and here

File: src/Processors/QtiV2/Out/QuestionTypes/PlaintextMapper.php [hunk 6]
```diff
@@ -13,19 +14,36 @@ public function convert(BaseQuestionType $questionType, $interactionIdentifier,
...
+            $feedbackOptions['genral_feedback'] = $metadata->get_distractor_rationale();
+        }
+        
         $interaction = new ExtendedTextInteraction($interactionIdentifier);
         $interaction->setLabel($interactionLabel);
         $interaction->setPrompt($this->convertStimulusForPrompt($question->get_stimulus()));
         $interaction->setFormat(TextFormat::PLAIN);
         $interaction->setMinStrings(1);
-        $interaction->setMaxStrings(1);
+        if(isset($questionData['max_length'])){
```
Comment (jack-lin-lrn) [inline src/Processors/QtiV2/Out/QuestionTypes/PlaintextMapper.php]: Can we have spaces beside `if condition`?

File: src/Processors/QtiV2/Out/QuestionTypes/PlaintextMapper.php [hunk 7]
```diff
@@ -13,19 +14,36 @@ public function convert(BaseQuestionType $questionType, $interactionIdentifier,
...
+            $interaction->setExpectedLength($questionData['max_length']);
+            $interaction->setMaxStrings($questionData['max_length']);
+        }else{
```
Comment (jack-lin-lrn) [inline src/Processors/QtiV2/Out/QuestionTypes/PlaintextMapper.php]: need spaces beside `else`

File: src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php [hunk 8]
```diff
@@ -31,15 +31,41 @@ public function buildValidation($responseIdentifier, $validation, $isCaseSensiti
...
+        $score = 0;
+        $maxscore = 0;
+        $penalty = 0;
+        
+        if (method_exists($validation, 'get_valid_response')) {
             $score = $validation->get_valid_response()->get_score();
-            $responseProcessing = QtiResponseProcessingBuilder::build($score);
+        }
+
+        if (method_exists($validation, 'get_max_score') && $validation->get_max_score()!='') {
```
Comment (jack-lin-lrn) [inline src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php]: Can we have spaces beside `!=`?

File: src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php [hunk 9]
```diff
@@ -31,15 +31,41 @@ public function buildValidation($responseIdentifier, $validation, $isCaseSensiti
...
+            $type[] = 'maxscore';
+            $maxscore = $validation->get_max_score();
+        }
+        if (method_exists($validation, 'get_penalty') && $validation->get_penalty()!='') {
```
Comment (jack-lin-lrn) [inline src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php]: need spaces beside `!=`

File: src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php [hunk 10]
```diff
@@ -31,15 +31,41 @@ public function buildValidation($responseIdentifier, $validation, $isCaseSensiti
...
+            $type[] = 'penalty';
+            $penalty = $validation->get_penalty();
+        }
+
+        if (sizeof($responseIdentifiers)>1) {
```
Comment (jack-lin-lrn) [inline src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php]: need spaces beside `>`

File: src/Processors/QtiV2/Out/QuestionTypes/LongtextMapper.php (hunk 1 above)
Comment (pooja31chaudhary) [reply src/Processors/QtiV2/Out/QuestionTypes/LongtextMapper.php]: Done

File: src/Processors/QtiV2/Out/QuestionTypes/LongtextV2Mapper.php (hunk 2 above)
Comment (pooja31chaudhary) [reply src/Processors/QtiV2/Out/QuestionTypes/LongtextV2Mapper.php]: Done

File: src/Processors/QtiV2/Out/QuestionTypes/PlaintextMapper.php (hunk 3 above)
Comment (pooja31chaudhary) [reply src/Processors/QtiV2/Out/QuestionTypes/PlaintextMapper.php]: Done

File: src/Processors/QtiV2/Out/ResponseProcessing/QtiResponseProcessingBuilder.php (hunk 4 above)
Comment (pooja31chaudhary) [reply src/Processors/QtiV2/Out/ResponseProcessing/QtiResponseProcessingBuilder.php]: Done

File: src/Processors/QtiV2/Out/Validation/PlaintextValidationBuilder.php (hunk 5 above)
Comment (pooja31chaudhary) [reply src/Processors/QtiV2/Out/Validation/PlaintextValidationBuilder.php]: Done

File: src/Processors/QtiV2/Out/QuestionTypes/PlaintextMapper.php (hunk 6 above)
Comment (pooja31chaudhary) [reply src/Processors/QtiV2/Out/QuestionTypes/PlaintextMapper.php]: Done

File: src/Processors/QtiV2/Out/QuestionTypes/PlaintextMapper.php (hunk 7 above)
Comment (pooja31chaudhary) [reply src/Processors/QtiV2/Out/QuestionTypes/PlaintextMapper.php]: Done

File: src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php (hunk 8 above)
Comment (pooja31chaudhary) [reply src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php]: Done

File: src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php (hunk 9 above)
Comment (pooja31chaudhary) [reply src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php]: Done

File: src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php (hunk 10 above)
Comment (pooja31chaudhary) [reply src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php]: Done

File: src/Processors/QtiV2/Out/Validation/LongtextValidationBuilder.php [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+        $responseDeclaration->setCardinality(Cardinality::SINGLE);
+        $responseDeclaration->setBaseType(BaseType::STRING);
+
+        return $responseDeclaration;
+    }
+
+    public function buildValidation($responseIdentifier, $validation, $isCaseSensitive = true, $distractorRationaleResponseLevel = array())
+    {
+        $responseProcessing = null;
+        if (empty($validation) && (method_exists($validation, 'get_max_score')) && (empty($validation->get_max_score()))) {
```
Comment (jack-lin-lrn) [inline src/Processors/QtiV2/Out/Validation/LongtextValidationBuilder.php]: Can we remove redundant outer parentheses?
```
if (empty($validation) && method_exists($validation, 'get_max_score') && empty($validation->get_max_score())) {
```

File: src/Processors/QtiV2/Out/Validation/PlaintextValidationBuilder.php [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+        $responseDeclaration->setCardinality(Cardinality::SINGLE);
+        $responseDeclaration->setBaseType(BaseType::STRING);
+
+        return $responseDeclaration;
+    }
+
+    public function buildValidation($responseIdentifier, $validation, $isCaseSensitive = true, $distractorRationaleResponseLevel = array())
+    {
+        $responseProcessing = null;
+        if (empty($validation) && (method_exists($validation, 'get_max_score')) && (empty($validation->get_max_score()))) {
```
Comment (jack-lin-lrn) [inline src/Processors/QtiV2/Out/Validation/PlaintextValidationBuilder.php]: same here, please

File: src/Processors/QtiV2/Out/AssessmentItemBuilder.php [hunk 13]
```diff
@@ -47,12 +45,8 @@ public function build($itemIdentifier, $itemLabel, array $questions, $content =
         $responseProcessingTemplates = [];
         foreach ($questions as $question) {
             $questionData = $question->to_array();
-            if (isset($questionData['data']['validation']['valid_response']['score'])) {
-                $score = $questionData['data']['validation']['valid_response']['score'];
-                $assessmentItem->setOutcomeDeclarations($this->buildOutcomeDeclarations($score));
-            } else {
-                $assessmentItem->setOutcomeDeclarations($this->buildOutcomeDeclarations(0));
-            }
+            $assessmentItem->setOutcomeDeclarations($this->buildOutcomeDeclarations(0));
```
Comment (jack-lin-lrn) [inline src/Processors/QtiV2/Out/AssessmentItemBuilder.php]: As param has already been remove from buildOutcomeDeclarations, is there any reason to keep 0 as param?

File: src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php [hunk 14]
```diff
@@ -31,15 +31,41 @@ public function buildValidation($responseIdentifier, $validation, $isCaseSensiti
...
+        $score = 0;
+        $maxscore = 0;
+        $penalty = 0;
+        
+        if (method_exists($validation, 'get_valid_response')) {
             $score = $validation->get_valid_response()->get_score();
-            $responseProcessing = QtiResponseProcessingBuilder::build($score);
+        }
+
+        if (method_exists($validation, 'get_max_score') && $validation->get_max_score() !='') {
```
[the remark this answers is not in the document]
Comment (pooja31chaudhary) [reply src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php]: Hi @jack-lin-lrn , according to PSR-2 format standards we can not add a space before closing perantheses.

File: src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php [hunk 15]
```diff
@@ -31,15 +31,41 @@ public function buildValidation($responseIdentifier, $validation, $isCaseSensiti
...
+            $type[] = 'maxscore';
+            $maxscore = $validation->get_max_score();
+        }
+        if (method_exists($validation, 'get_penalty') && $validation->get_penalty() !='') {
```
[the remark this answers is not in the document]
Comment (pooja31chaudhary) [reply src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php]: same as above.

File: src/Processors/QtiV2/Out/Validation/LongtextValidationBuilder.php (hunk 11 above)
Comment (pooja31chaudhary) [reply src/Processors/QtiV2/Out/Validation/LongtextValidationBuilder.php]: Done

File: src/Processors/QtiV2/Out/Validation/PlaintextValidationBuilder.php (hunk 12 above)
Comment (pooja31chaudhary) [reply src/Processors/QtiV2/Out/Validation/PlaintextValidationBuilder.php]: Done

File: src/Processors/QtiV2/Out/AssessmentItemBuilder.php (hunk 13 above)
Comment (pooja31chaudhary) [reply src/Processors/QtiV2/Out/AssessmentItemBuilder.php]: Its look like I forgot to remove this 0. Removed now.

File: src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php (hunk 14 above)
[the remark this answers is not in the document]
Comment (jack-lin-lrn) [reply src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php]: I totally agree with you. I mean we should add space before `''` not after. Now, it looks like there is no space between `!=` and `''`.

File: src/Processors/QtiV2/Out/Validation/LongtextValidationBuilder.php (hunk 11 above)
Comment (jack-lin-lrn) [reply src/Processors/QtiV2/Out/Validation/LongtextValidationBuilder.php]: Hi @pooja31chaudhary ,
Have you pushed the fixes?

**Comment by jack-lin-lrn** [comment]
Hi @pooja31chaudhary ,
Just as a reminder, there are few suggestions on the list.

File: src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php (hunk 14 above)
[the remark this answers is not in the document]
Comment (jack-lin-lrn) [reply src/Processors/QtiV2/Out/Validation/AbstractQuestionValidationBuilder.php]: Sorry for my poor description. You probably misunderstood it. 
```
$validation->get_max_score() !='') {         <- now, not expected
$validation->get_max_score() != '') {        <- expected
```
Could you add a space before empty string?

**Comment by rakeshqdsa** [comment]
This PRs code has been merged in PR#50 as most of the code was duplicate .

**Comment by ttton** [comment]
@pooja31chaudhary I'm wondering what the state of this PR is. It seems to have gone stale and hasn't been touched in a few weeks. Can I confirm that all fixes in this PR have been included elsewhere?

**Comment by rakeshqdsa** [comment]
Yes this is old PR and all the changes has been merged on PR#50 which is already merged on integration branch .

**Comment by ttton** [comment]
Closing this as duplicate.

**Pull request closed without merging**